org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/codeactions/LSPCodeActionQuickAssistProcessor.java

```diff
@@ -158,22 +151,23 @@ public ICompletionProposal[] computeQuickAssistProposals(IQuickAssistInvocationC
 		}
 
 		// Get the codeActions
-		CodeActionParams params = prepareCodeActionParams(infos, invocationContext.getOffset(), invocationContext.getLength());
+		CodeActionParams params = prepareCodeActionParams(document, invocationContext.getOffset(), invocationContext.getLength());
 		List<CompletableFuture<Void>> futures = Collections.emptyList();
 		try {
 			// Prevent infinite re-entrance by only computing proposals if there aren't any
 			if (!proposalsRefreshInProgress) {
 				proposals.clear();
-				futures = infos.stream()
-						.map(info -> info.getInitializedLanguageClient()
-								.thenComposeAsync(ls -> ls.getTextDocumentService().codeAction(params)
-										.thenAcceptAsync(actions -> actions.stream().filter(Objects::nonNull)
-												.map(action -> new CodeActionCompletionProposal(action, info))
-												.forEach(p -> processNewProposal(invocationContext, p)))))
-								.toList();
+				// Start all the servers computing actions - each server will append any code actions to the ongoing list of proposals
+				// as a side effect of this request
+				futures = executor.computeAll((w, ls) -> ls.getTextDocumentService()
```

Comment:
could we not use here collectAll and then save as the trouble of calling 
```
CompletableFuture<?> aggregateFutures = CompletableFuture
						.allOf(futures.toArray(new CompletableFuture[futures.size()]))
```
right after?

Reply:
No, because the intermediate (per-LS) futures are needed at line 178.

The logic for this function is not obvious, but this is one example (of many) where we have to compromise because we have to couple a 'push' (the asynchronous calls to the LS, which complete and then push their results to whatever next steps are enchained) to a 'pull' - where the UI is calling a method and expecting the result synchronously.

The way this method works is to start a load of calls to the possible servers in the foreground. We're not interested in a result from those CompletableFutures - they have `void` return type and do their work as a mutating side-effect, adding their intermediate results into a field on the class.

Their individual futures are combined into a single one that completes when they all do. If this single future completes before the timeout, then we're good: the whole result is available and can be returned synchronously from this method. If not, then this method will return the result so far, and chain follow-up actions onto the individual uncompleted futures so that they trigger a refresh.

It's a bit convoluted, but I have tried to be faithful to the original behaviour and do a straight port to the new API without attempting to improve or alter the logic. If we used `collectAll` as you suggest then we wouldn't be able to do multiple updates if we got several more results at separate times after the initial timeout. We could make the change you suggest but in that case the user would get the initial result up to that point, but would not see an update until *all* the remaining results had arrived. That might be sufficient but I have confined myself to porting the code and not trying to change the behaviour.